fix FSS mask loss claculation issue

jni2000 opened the pull request:
This PR is for fixing the issue as described in #652.

nasrinivasa commented on src/harness/reference_models/interference/interference.py:
```diff
@@ -413,45 +413,67 @@ def computeInterferenceFssCochannel(cbsd_grant, constraint, fss_info, max_eirp):
+   # Calculate the total of 1MHZ frequencies present in the grant frequency range
+   num_one_mhz_bin_seg1 = cbsd_freq_range/int(MHZ)
+   # Calculate the fss_mask for every 1MHZ frequency 
+   for i in range (0, num_one_mhz_bin_seg1 - 1):
```
Range automatically takes care of this as it generates numbers from 0 to num_one_mhz_bin_seg1-1

jni2000 replied:
Updated   the Range usage

nasrinivasa commented on src/harness/reference_models/interference/interference.py:
```diff
@@ -413,45 +413,67 @@ def computeInterferenceFssCochannel(cbsd_grant, constraint, fss_info, max_eirp):
+   num_one_mhz_bin_seg1 = cbsd_freq_range/int(MHZ)
+   # Calculate the fss_mask for every 1MHZ frequency 
+   for i in range (0, num_one_mhz_bin_seg1 - 1):
+     fss_mask_attenuation += 10 **((dbToLinear(-(constraint.high_frequency - cbsd_grant.low_frequency)  - 0.5 -i)*0.6 - 0.5)/10) 
```
The braces don't look right. I think it should be as follows:
10 **(dbToLinear(-((constraint.high_frequency - cbsd_grant.low_frequency)  - 0.5 -i)*0.6 - 0.5)/10)

jni2000 replied:
Updated   as suggested in the comments, also updated the equation removed 'dbToLinear'   conversion

nasrinivasa commented on src/harness/reference_models/interference/interference.py:
```diff
@@ -413,45 +413,67 @@ def computeInterferenceFssCochannel(cbsd_grant, constraint, fss_info, max_eirp):
+    # Calculate the fss_mask for every 1MHZ frequency
+    for i in range (0, num_one_mhz_bin_seg2 - 1):
+      fss_mask_attenuation += 10 **((dbToLinear(-(offset - cbsd_grant.low_frequency)  - 0.5 -i)*0.25 - 30.5)/10)
+    # Calculate the average attenuation value per MHZ
```
Same comments as above for the segment calculation: use range (0,num_one_mhz_bin_seg2 ) and the braces do not look right

jni2000 commented on src/harness/reference_models/interference/interference.py:
```diff
@@ -409,46 +409,67 @@ def computeInterferenceFssCochannel(cbsd_grant, constraint, fss_info, max_eirp):
+  
+  # If the grant overlap frequency lies between 3650 to 3700 MHz
+  # Segment 1 calculation is applied
+  if ( offset <= cbsd_grant.low_frequency and
```
(constraint.low_frequency,   constraint.high_frequency) is the FSS    frequency range for which we need to do blocking protection.     constraint.low_frequency is 3550 constraint.high_frequency is 3700

jni2000 commented on src/harness/reference_models/interference/interference.py:
```diff
@@ -409,46 +409,67 @@ def computeInterferenceFssCochannel(cbsd_grant, constraint, fss_info, max_eirp):
 def getFssMaskLoss(cbsd_grant, constraint):
```
(constraint.low_frequency,   constraint.high_frequency) is the FSS    frequency range for which we need to do blocking protection

jni2000 commented on src/harness/reference_models/interference/interference.py:
```diff
@@ -409,46 +409,67 @@ def computeInterferenceFssCochannel(cbsd_grant, constraint, fss_info, max_eirp):
+   for i in range (0, num_one_mhz_bin_seg1 - 1):
+     fss_mask_attenuation += 10 **((dbToLinear(-(constraint.high_frequency - cbsd_grant.low_frequency)  - 0.5 -i)*0.6 - 0.5)/10) 
+
+   # Calculate the average attenuation value per MHZ
```
Corrected   the indentation issues

jni2000 commented on src/harness/reference_models/interference/interference.py:
```diff
@@ -409,46 +409,67 @@ def computeInterferenceFssCochannel(cbsd_grant, constraint, fss_info, max_eirp):
+    fss_mask_attenuation_seg2 = linearToDb(fss_mask_attenuation/num_one_mhz_bin_seg2)
+ 
+  else:
+    print "Not in the Fss Filter Mask Range"
```
Added   exception

jni2000 commented on src/harness/reference_models/interference/interference.py:
```diff
@@ -409,46 +409,67 @@ def computeInterferenceFssCochannel(cbsd_grant, constraint, fss_info, max_eirp):
+  fss_mask_attenuation_seg1 = 0
+  fss_mask_attenuation_seg2 = 0
+  
+  # If the grant overlap frequency lies between 3650 to 3700 MHz
```
constraint.low_frequency   3550 constraint.high_frequency  3700 , the constraint frequency is he   FSS  frequency range for which we need   to do blocking protection

jni2000 commented on src/harness/reference_models/interference/interference.py:
```diff
@@ -409,46 +409,67 @@ def computeInterferenceFssCochannel(cbsd_grant, constraint, fss_info, max_eirp):
+   num_one_mhz_bin_seg1 = cbsd_freq_range/int(MHZ)
+   # Calculate the fss_mask for every 1MHZ frequency 
+   for i in range (0, num_one_mhz_bin_seg1 - 1):
+     fss_mask_attenuation += 10 **((dbToLinear(-(constraint.high_frequency - cbsd_grant.low_frequency)  - 0.5 -i)*0.6 - 0.5)/10) 
```
Please see explanation in the new comment we added.

jni2000 commented:
Some explanation on the formula

```
OffsetFreqMHz = fp1 (MHz) - 50 MHz
MaskSeg1 defined as fp1-50MHz to fp1
MaskSeg2 defined as fp1-150MHz to fp1-50MHz
fss_mask_attenuation = 0
if grant freqs in MaskSeg1
   NumOneMHzBinsSeg1 = (overlappedGrantMaskSeg1_MaxFreqMHz - overlappedGrantMaskSeg1_MinFreqMHz)/MHz
        for i = 0:NumOneMHzBinsSeg1-1
            fss_mask_attenuation +=10^[-(fp1 - overlappedCbsdMaskSeg1_MinFreqMHz - 0.5 -i)*0.6 - 0.5]/10
        endfor

    fss_mask_attenuation_seg1 = fss_mask_attenuation/NumOneMHzBinsSeg1
endif

fss_mask_attenuation = 0
if grant freqs in MaskSeg2
   NumOneMHzBinsSeg2 = (overlappedGrantMaskSeg2_MaxFreqMHz - overlappedGrantMaskSeg2_MinFreqMHz)/MHz
        for i = 0:NumOneMHzBinsSeg2-1
            fss_mask_attenuation +=10^[-(OffsetFreqMHz - overlappedCbsdMaskSeg2_MinFreqMHz - 0.5 -i)*0.25 - 30.5]/10
        endfor

    fss_mask_attenuation_seg2 = fss_mask_attenuation/NumOneMHzBinsSeg2
endif

fss_mask_attenuation_total = fss_mask_attenuation_seg1 + fss_mask_attenuation_seg2

```
Example 1 covering MaskSeg1

Grant is exactly 50 MHz extending up to fp1, so overlapping the dashed blue segment. In this case, filter mask attenuation (in dB) is -30.2, -29.6, ..., -0.8 at the 1 MHz centers.  Values obtained by solving: 

-(50-0.5-[0:49])*0.6-0.5

Example 2 covering MaskSeg2

fp1: 3700 MHz
Offset: 3650 MHz
MaskSeg1: 3650 to 3700 MHz
MaskSeg2: 3550 to 3650 MHz
Grant: 3550 – 3560 MHz

At fp1-150 MHz (3550 in this case) mask loss is -55.5 dB (-100*0.25 - 50*0.6 - 0.5)
In this case, filter mask attenuation values (in dB) starting at 3550.5 MHz are -55.375, ..., -53.125 (last one at 3559.5 MHz)
at the 1 MHz centers.  Values obtained by solving:
-(100 - 0.5 -[0:9])*0.25 - 30.5]

jni2000 commented on src/harness/reference_models/interference/interference.py:
```diff
@@ -409,46 +409,67 @@ def computeInterferenceFssCochannel(cbsd_grant, constraint, fss_info, max_eirp):
+  # If the grant frequency overlaps both the segments, the overlapping
+  # range is considered and calculation is done on both segments
+  elif ((offset - 100*MHZ) <= cbsd_grant.low_frequency and
+        constraint.high_frequency >= cbsd_grant.high_frequency):
```
Here   considered constraint.high_frequency is    3700,  so we are checking the   grant frequencies with constraint.high_frequency.

jni2000 commented on src/harness/reference_models/interference/interference.py:
```diff
@@ -409,46 +409,67 @@ def computeInterferenceFssCochannel(cbsd_grant, constraint, fss_info, max_eirp):
+          offset >= cbsd_grant.high_frequency):
+    fss_mask_attenuation = 0
+    # Calculate the total of 1MHZ frequencies present in the grant frequency range
+    num_one_mhz_bin_seg2 = cbsd_freq_range/int(MHZ)
```
Here we   are considering grants frequency is always less than or equal to 3700, so   grants will be always present in FSS blocking frequencies ( 3550 - 3700),   this part is taken care of.

nasrinivasa commented on src/harness/reference_models/interference/interference.py:
```diff
@@ -413,57 +394,73 @@ def computeInterferenceFssCochannel(cbsd_grant, constraint, fss_info, max_eirp):
-
-  return fss_mask_loss
+  else:
+    raise Exception("Fss Mask is not calculated for grants falling within band 3700 to 4200")
```
Are we handling the exception that is raised here in the calling function?

piotrfilipiuk commented on src/harness/reference_models/interference/interference.py:
```diff
@@ -413,57 +394,73 @@ def computeInterferenceFssCochannel(cbsd_grant, constraint, fss_info, max_eirp):
+    num_one_mhz_bin_seg1 = int((cbsd_grant.high_frequency - offset)/MHZ)
+    fss_mask_attenuation = 0
+    for i in range (0, num_one_mhz_bin_seg1):
+      fss_mask_attenuation +=10**(-(((constraint.high_frequency - offset)/int(MHZ)) - 0.5 -i)*0.6 - 0.5)/10
```
This code is duplicated, please refactor. Similarly other places where fss_mask_attenuation is being calculated.

jni2000 replied:
can you please be more specific on what shall be done ?

piotrfilipiuk commented on src/harness/reference_models/interference/interference.py:
```diff
@@ -413,57 +394,73 @@ def computeInterferenceFssCochannel(cbsd_grant, constraint, fss_info, max_eirp):
+    for i in range (0, num_one_mhz_bin_seg1 ):
+      fss_mask_attenuation +=10**(-(((constraint.high_frequency - cbsd_grant.low_frequency)/int(MHZ)) - 0.5 -i)*0.6 - 0.5)/10
+    # Calculate the average attenuation value per MHZ
+    fss_mask_attenuation_seg1 = fss_mask_attenuation/num_one_mhz_bin_seg1
```
I am not sure why fss_mask_attenuation is divided by the channel size, unless the function returns loss per MHz. If that is the case, please document the function. However, since for blocking protection we need to use total amount of power (not per MHz), I do not see a reason to calculate loss per MHz.

jni2000 replied:
We are calculating attenuation due to FSS filter in the center of each 1 MHz bin and average it over the bandwidth of grant. The per MHz computation is to only take into account filter loss at each MHz bin but is then aggregated over grant bandwidth. At the end we get FSS filter's contribution to the grant bandwidth and not on a per MHz basis.

sbdt-google commented on src/harness/reference_models/interference/interference.py:
```diff
@@ -413,57 +394,73 @@ def computeInterferenceFssCochannel(cbsd_grant, constraint, fss_info, max_eirp):
+    num_one_mhz_bin_seg1 = cbsd_freq_range/int(MHZ)
+    # Calculate the fss_mask for every 1MHZ frequency
+    for i in range (0, num_one_mhz_bin_seg1 ):
+      fss_mask_attenuation +=10**(-(((constraint.high_frequency - cbsd_grant.low_frequency)/int(MHZ)) - 0.5 -i)*0.6 - 0.5)/10
```
For the db<-> linear conversion, please use instead: 
```
  dbToLinear()
  linearToDb0
```
that are defined in this same file. This will avoid such errors

sbdt-google replied:
Also why do you use the strange construct :  /int(MHZ) here and below ??
If you want to take the integer part of the division, you should do:
  `int(a/MHZ)` instead of `a/int(MHZ)`. 
As the last one assumes that a is integer.

sbdt-google commented on src/harness/reference_models/interference/interference.py:
```diff
@@ -15,21 +15,16 @@
 ==================================================================================
   Compute Interference caused by a grant for all the incumbent types
   APIs in this file are used by IAP and Aggregate Interference Reference Models
-
```
Why are you removing in this file all empty lines which improves readability ?

jni2000 replied:
Added   back the empty lines

sbdt-google commented on src/harness/reference_models/interference/interference.py:
```diff
@@ -198,7 +189,6 @@ def getProtectedChannels(low_freq_hz, high_freq_hz):
 
 def findGrantsInsideNeighborhood(grants, protection_point, entity_type):
   """Finds grants inside protection entity neighborhood.
-
```
Again here and below, all empty lines are removed, although we always keep one empty line before the "Args" section for readability

jni2000 replied:
Added   the empty lines

sbdt-google commented on src/harness/reference_models/interference/interference.py:
```diff
@@ -413,57 +394,73 @@ def computeInterferenceFssCochannel(cbsd_grant, constraint, fss_info, max_eirp):
+  if ( offset <= cbsd_grant.low_frequency and
+         constraint.high_frequency >= cbsd_grant.high_frequency):
+    # Calculate the total of 1MHZ frequencies present in the grant frequency range
+    num_one_mhz_bin_seg1 = cbsd_freq_range/int(MHZ)
```
I guess you want the integer part, then use:
int(cbsd_freq_range/MHZ), and not the current line

Otherwise if cbsd_freq_range, it will still do a float point division giving back a floating point

jni2000 replied:
thanks.

sbdt-google commented on src/harness/reference_models/interference/interference.py:
```diff
@@ -413,57 +394,73 @@ def computeInterferenceFssCochannel(cbsd_grant, constraint, fss_info, max_eirp):
+          offset >= cbsd_grant.high_frequency):
+    fss_mask_attenuation = 0
+    # Calculate the total of 1MHZ frequencies present in the grant frequency range
+    num_one_mhz_bin_seg2 = cbsd_freq_range/int(MHZ)
```
Again the /int(MHZ) is not safe.
Please correct everywhere in the function

jni2000 replied:
Modified   formula to int(a/MHZ)

sbdt-google commented on src/harness/reference_models/interference/interference.py:
```diff
@@ -413,57 +394,73 @@ def computeInterferenceFssCochannel(cbsd_grant, constraint, fss_info, max_eirp):
+   
 def computeInterferenceFssBlocking(cbsd_grant, constraint, fss_info, max_eirp):
   """Computes interference that a grant causes to a FSS protection point.
-
```
Again why removing the blank lines here and everywhere else. This goes against the convention 
used in the test harness code.

jni2000 replied:
Added   back the removed new lines

kate-harrison commented:
James, we really need an update on this ASAP if we want to have any chance of meeting our Tuesday deadline. When can you send the next update?

jni2000 commented on src/harness/reference_models/interference/interference.py:
```diff
@@ -179,10 +172,8 @@ def linearToDb(x):
   Performs 5MHz IAP channelization and returns a list of tuple containing
-  (low_freq,high_freq)
-
+  ,(low_freq,high_freq)
```
Removed   the comma

jni2000 commented:
Serge and Piotr

We have addressed most of your review comments and are working on your following comments:

1. "This   code is duplicated, please refactor. Similarly other places where   fss_mask_attenuation is being calculated.", can you please be more specific on what is duplicate ?
2. "I am   not sure why fss_mask_attenuation is divided by the channel size, unless the   function returns loss per MHz. If that is the case, please document the   function. However, since for blocking protection we need to use total amount   of power (not per MHz), I do not see a reason to calculate loss per MHz.", we are discussing this with our RF engineers and get back to you on this one soon.

nasrinivasa commented:
James, 
   I am ok with the changes that you made. 

Navin

jni2000 commented:
Thanks, Navin. Serge and Piotr, please take a look of our responses and changes. Thanks.

sbdt-google commented on src/harness/reference_models/interference/interference.py:
```diff
@@ -411,50 +412,82 @@ def computeInterferenceFssCochannel(cbsd_grant, constraint, fss_info, max_eirp):
+    num_one_mhz_bin_seg2 = int((offset - cbsd_grant.low_frequency)/MHZ)
+    for i in range (0, num_one_mhz_bin_seg2):
+      fss_mask_attenuation += dbToLinear((-(num_one_mhz_bin_seg2) - 0.5 -i)*0.25 - 30.5)
+    fss_mask_attenuation_seg2 = fss_mask_attenuation/num_one_mhz_bin_seg2
```
This logic is wrong, you cannot add fss_mask_attenuation_seg1 and seg2 like that after having done the average. It needs to be added before the division.

This whole case needs can take the exampled of the first 2 cases. But instead of that, please check my last comment which fixes the whole function.

jni2000 replied:
Each segment's average is independent of other. The segments are broken and contribution is independently accounted.

jni2000 commented:
Thanks, Serge. We are assessing your comments and will get back soon. Thanks.

jni2000 commented on src/harness/reference_models/interference/interference.py:
```diff
@@ -411,50 +412,82 @@ def computeInterferenceFssCochannel(cbsd_grant, constraint, fss_info, max_eirp):
+  """
+  # Check whether the grant frequency lies within the Fss Blocking protection frequency
+  assert cbsd_grant.low_frequency >= constraint.low_frequency, 'Grant Low frequency is less than Fss protection Low frequency'
+  assert cbsd_grant.high_frequency <= constraint.high_frequency, 'Grant High frequency is greater than Fss protection High frequency'
```
Awais: The logic is correct. as
assert cbsd_grant.low_frequency < constraint.low_frequency ---- 'Grant Low frequency is less than Fss protection Low frequency'
assert cbsd_grant.high_frequency <= constraint.low_frequency  ---- 'Grant High frequency is less than or equal to Fss passband low frequency'

jni2000 commented on src/harness/reference_models/interference/interference.py:
```diff
@@ -411,50 +412,82 @@ def computeInterferenceFssCochannel(cbsd_grant, constraint, fss_info, max_eirp):
+    num_one_mhz_bin_seg1 = int(cbsd_freq_range/MHZ)
+    # Calculate the fss_mask for every 1MHZ frequency
+    for i in range (0, num_one_mhz_bin_seg1):
+      fss_mask_attenuation +=dbToLinear(-(int((constraint.high_frequency - cbsd_grant.low_frequency)/MHZ) - 0.5 -i)*0.6 - 0.5)
```
We double checked and test run and we are sure the logic implemented is correct. As for the optimization you suggested, as well as all others related, we are in processing of assessing the accuracy. if not issues, we will implement. But for now, let's keep what we have due to the V&V urgency.

jni2000 commented on src/harness/reference_models/interference/interference.py:
```diff
@@ -411,50 +412,82 @@ def computeInterferenceFssCochannel(cbsd_grant, constraint, fss_info, max_eirp):
+    for i in range (0, num_one_mhz_bin_seg1):
+      fss_mask_attenuation +=dbToLinear(-(int((constraint.high_frequency - cbsd_grant.low_frequency)/MHZ) - 0.5 -i)*0.6 - 0.5)
+    # Calculate the average attenuation value per MHZ
+    fss_mask_attenuation_seg1 = fss_mask_attenuation/num_one_mhz_bin_seg1
```
We double checked and test run and we are sure the logic implemented is correct. As for the optimization you suggested, as well as all others related, we are in processing of assessing the accuracy. if not issues, we will implement. But for now, let's keep what we have due to the V&V urgency.

jni2000 commented on src/harness/reference_models/interference/interference.py:
```diff
@@ -411,50 +412,82 @@ def computeInterferenceFssCochannel(cbsd_grant, constraint, fss_info, max_eirp):
+    for i in range (0, num_one_mhz_bin_seg2):
+      fss_mask_attenuation += dbToLinear(-(int((offset - cbsd_grant.low_frequency)/MHZ) - 0.5 -i)*0.25 - 30.5)
+    # Calculate the average attenuation value per MHZ
+    fss_mask_attenuation_seg2 = fss_mask_attenuation/num_one_mhz_bin_seg2
```
We double checked and test run and we are sure the logic implemented is correct. As for the optimization you suggested, as well as all others related, we are in processing of assessing the accuracy. if not issues, we will implement. But for now, let's keep what we have due to the V&V urgency.

jni2000 commented on src/harness/reference_models/interference/interference.py:
```diff
@@ -411,50 +412,82 @@ def computeInterferenceFssCochannel(cbsd_grant, constraint, fss_info, max_eirp):
+
+  # If the grant frequency overlaps both the segments, the overlapping
+  # range is considered and calculation is done on both segments
+  elif ((offset - 100*MHZ) <= cbsd_grant.low_frequency and
```
We are not looking for overlap with the passband rather with filter roll-off segments. The logic is correct.

jni2000 commented on src/harness/reference_models/interference/interference.py:
```diff
@@ -411,50 +412,82 @@ def computeInterferenceFssCochannel(cbsd_grant, constraint, fss_info, max_eirp):
+    for i in range (0, num_one_mhz_bin_seg2):
+      fss_mask_attenuation += dbToLinear((-(num_one_mhz_bin_seg2) - 0.5 -i)*0.25 - 30.5)
+    fss_mask_attenuation_seg2 = fss_mask_attenuation/num_one_mhz_bin_seg2
 
```
We double checked and test run and we are sure the logic implemented is correct. As for the optimization you suggested, as well as all others related, we are in processing of assessing the accuracy. if not issues, we will implement. But for now, let's keep what we have due to the V&V urgency.

jni2000 commented on src/harness/reference_models/interference/interference.py:
```diff
@@ -411,50 +412,82 @@ def computeInterferenceFssCochannel(cbsd_grant, constraint, fss_info, max_eirp):
+
+  # If the grant overlap frequency lies between fp1-150 MHz to fp1-50, where fp1 is the starting frequency of FSS Passband
+  # Segment 2 calculation is applied
+  elif ((offset - 100*MHZ) <= cbsd_grant.low_frequency and
```
can you explain the case ?

sbdt-google commented:
I am sorry I cannot approve this PR. It is still has bugs, despite what you are saying (in addition of being difficult to understand/analyze).

Because of the "V&V urgency" as you put it, I don't have time to reexplain to the bugs in your code a second time. So let's delegate to Kate what should be done.

kate-harrison commented:
I have discussed with Serge offline and he has generously offered to prepare another PR ASAP to address the concerns he has raised in this PR. Please review that one and let him know if you have any concerns. If people are comfortable, we can use that one instead of this PR. Unfortunately we need to pick one of them by the end of the day so that we can meet our deadline.

jni2000 commented:
This works for us. I would like to connect Serge and Awais (our side engineer) to discuss this matter.

jcereced commented:
can you guys please also include Navin in this discussions?

jcereced commented:
sounds good...I misunderstood that you guys were having a side chat to discuss things

jni2000 commented:
Javi, you did not misunderstand. I did proposed to discuss offline, but suggested by Kate and Serge otherwise. Good is that you and Navin did not miss anything yet.

sbdt-google commented:
Please check the alternative PR #702. 
I introduced also has proper unit tests, all failing in your proposed implementation.

jni2000 commented:
taken care by PR #702
